ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,304 @@
+   * @param numSegments the total number of segments that the log abstraction will contain.
+   * @throws IOException if there is an I/O error creating/loading the segment files or creating the {@link LogSegment}
+   * instances
+   */
+  private void loadSegments(long segmentCapacity, long numSegments)
+      throws IOException {
+    File dir = new File(dataDir);
+    File[] segmentFiles = dir.listFiles(LogSegmentNameHelper.LOG_FILE_FILTER);
+    if (segmentFiles == null || segmentFiles.length == 0) {
+      String firstSegmentName = LogSegmentNameHelper.generateFirstSegmentName(numSegments);
+      File segmentFile = allocate(LogSegmentNameHelper.nameToFilename(firstSegmentName), segmentCapacity);
+      // to be backwards compatible, headers are not written for a log segment if it is the only log segment.
+      LogSegment segment = new LogSegment(firstSegmentName, segmentFile, segmentCapacity, metrics, numSegments > 1);
+      segmentsByName.put(segment.getName(), segment);
+    } else {
+      for (File segmentFile : segmentFiles) {
+        String name = LogSegmentNameHelper.nameFromFilename(segmentFile.getName());
+        LogSegment segment;
+        if (numSegments == 1) {
+          // for backwards compatibility, a single segment log is loaded by providing capacity since the old logs have
+          // no headers
+          segment = new LogSegment(name, segmentFile, segmentCapacity, metrics, false);
+        } else {
+          segment = new LogSegment(name, segmentFile, metrics);
+        }
+        segmentsByName.put(segment.getName(), segment);
+        remainingUnallocatedCapacity.addAndGet(-segmentCapacity);
+      }
     }
-    logger.trace("Log : {} bytes written to the log from read channel at {}, bytesWritten: {}", file.getAbsolutePath(),
-        offset, bytesWritten);
+    activeSegment = segmentsByName.firstEntry().getValue();
   }
 
   /**
-   * Close this log
+   * Allocates a file named {@code filename} and of capacity {@code size}.
+   * @param filename the intended filename of the file.
+   * @param size the intended size of the file.
+   * @return a {@link File} instance that points to the created file named {@code filename} and capacity {@code size}.
+   * @throws IOException if the there is any I/O error in allocating the file.
    */
-  void close()
+  private File allocate(String filename, long size)
       throws IOException {
-    fileChannel.close();
+    // TODO (DiskManager changes): This is intended to "request" the segment file from the DiskManager which will have
+    // TODO (DiskManager changes): a pool of segments.
+    File segmentFile = new File(dataDir, filename);
+    Utils.preAllocateFileIfNeeded(segmentFile, size);
+    remainingUnallocatedCapacity.addAndGet(-size);
+    return segmentFile;
   }
 
-  public void flush()
+  /**
+   * Rolls the log segment over if required. Also ensures enough capacity.
```

Comment:
Changed.
Ensures capacity by calling `ensureCapacity()`. Is that what your question is about?